core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -343,17 +343,31 @@ private[spark] class TaskSchedulerImpl(
         }
         taskIdToTaskSetManager.get(tid) match {
           case Some(taskSet) =>
+            var executorId: String = null
             if (TaskState.isFinished(state)) {
               taskIdToTaskSetManager.remove(tid)
               taskIdToExecutorId.remove(tid).foreach { execId =>
+                executorId = execId
                 if (executorIdToTaskCount.contains(execId)) {
                   executorIdToTaskCount(execId) -= 1
                 }
               }
             }
             if (state == TaskState.FINISHED) {
+              // In some case, executor has already removed by driver for heartbeats timeout, but
+              // at sometime, before executor killed  by cluster, the task of running on this
+              // executor is finished and return task success state to driver. However, this kinds
+              // of task should be ignored, because the task on this executor is already re-queued
+              // by driver. For more details, can check in SPARK-14485.
               taskSet.removeRunningTask(tid)
-              taskResultGetter.enqueueSuccessfulTask(taskSet, tid, serializedData)
+              if (executorId != null && !executorIdToTaskCount.contains(executorId)) {
+                logInfo(
+                  ("Ignoring update with state %s for TID %s because its executor has already " +
```

Comment:
There are lots of places where we could use interpolation instead of format strings. But there's no need to go around and modify existing, working code, let's try first to just enforce that rule for new code.